Mathlib/RingTheory/TensorProduct/MvPolynomial.lean

```diff
@@ -0,0 +1,169 @@
+
+variable   [AddCommMonoid N] [Module R N]
+
+/-- The tensor product of a polynomial ring by a module is
+  linearly equivalent to a Finsupp of a tensor product -/
+noncomputable def rTensor :
+    MvPolynomial σ S ⊗[R] N ≃ₗ[S] (σ →₀ ℕ) →₀ (S ⊗[R] N) :=
+  TensorProduct.finsuppLeft' _ _ _ _ _
+
+lemma rTensor_apply_tmul (p : MvPolynomial σ S) (n : N) :
+    rTensor (p ⊗ₜ[R] n) = p.sum (fun i m ↦ Finsupp.single i (m ⊗ₜ[R] n)) :=
+  TensorProduct.finsuppLeft_apply_tmul p n
+
+lemma rTensor_apply_tmul_apply (p : MvPolynomial σ S) (n : N) (d : σ →₀ ℕ) :
+    rTensor (p ⊗ₜ[R] n) d = (coeff d p) ⊗ₜ[R] n :=
+  TensorProduct.finsuppLeft_apply_tmul_apply p n d
+
+lemma rTensor_apply (t : MvPolynomial σ S ⊗[R] N) (d : σ →₀ ℕ) :
+    rTensor t d = ((lcoeff S d).restrictScalars R).rTensor N t :=
+  TensorProduct.finsuppLeft_apply t d
+
+lemma rTensor_symm_apply_single (d : σ →₀ ℕ) (s : S) (n : N) :
+    rTensor.symm (Finsupp.single d (s ⊗ₜ n)) =
+      (monomial d s) ⊗ₜ[R] n :=
+  TensorProduct.finsuppLeft_symm_apply_single d s n
+
+/-- The tensor product of the polynomial algebra by a module
+  is linearly equivalent to a Finsupp of that module -/
+noncomputable def scalarRTensor :
+    MvPolynomial σ R ⊗[R] N ≃ₗ[R] (σ →₀ ℕ) →₀ N :=
+  TensorProduct.finsuppScalarLeft _ _ _
+
+lemma scalarRTensor_apply_tmul (p : MvPolynomial σ R) (n : N) :
+    scalarRTensor (p ⊗ₜ[R] n) = p.sum (fun i m ↦ Finsupp.single i (m • n)) :=
+  TensorProduct.finsuppScalarLeft_apply_tmul p n
+
+lemma scalarRTensor_apply_tmul_apply (p : MvPolynomial σ R) (n : N) (d : σ →₀ ℕ):
+    scalarRTensor (p ⊗ₜ[R] n) d = (coeff d p) • n :=
+  TensorProduct.finsuppScalarLeft_apply_tmul_apply p n d
+
+lemma scalarRTensor_symm_apply_single (d : σ →₀ ℕ) (n : N) :
+    scalarRTensor.symm (Finsupp.single d n) = (monomial d 1) ⊗ₜ[R] n :=
+  TensorProduct.finsuppScalarLeft_symm_apply_single d n
+
+end Module
+
+section Algebra
+
+variable [CommSemiring N] [Algebra R N]
+
+/-- The algebra morphism from a tensor product of a polynomial algebra
+  by an algebra to a polynomial algebra -/
+noncomputable def rTensorAlgHom :
+    (MvPolynomial σ S) ⊗[R] N →ₐ[S] MvPolynomial σ (S ⊗[R] N) :=
+  Algebra.TensorProduct.lift
+    (mapAlgHom Algebra.TensorProduct.includeLeft)
+    ((IsScalarTower.toAlgHom R (S ⊗[R] N) _).comp Algebra.TensorProduct.includeRight)
+    (fun p n => by simp [commute_iff_eq, algebraMap_eq, mul_comm])
+
+lemma rTensorAlgHom_coeff_tmul
+    (p : MvPolynomial σ S) (n : N) (d : σ →₀ ℕ) :
+    coeff d (rTensorAlgHom (p ⊗ₜ[R] n)) = (coeff d p) ⊗ₜ[R] n := by
+  rw [rTensorAlgHom, Algebra.TensorProduct.lift_tmul]
+  rw [AlgHom.coe_comp, IsScalarTower.coe_toAlgHom', Function.comp_apply,
+    Algebra.TensorProduct.includeRight_apply]
+  rw [algebraMap_eq, mul_comm, coeff_C_mul]
+  simp [mapAlgHom, coeff_map]
+
+lemma rTensorAlgHom_toLinearMap :
+    (rTensorAlgHom :
+      MvPolynomial σ S ⊗[R] N →ₐ[S] MvPolynomial σ (S ⊗[R] N)).toLinearMap =
+      (finsuppLeft' _ _ _ _ _).toLinearMap := by
```

Comment:
Is it possible to use `rTensor` instead of `finsuppLeft'` here?

Reply:
Yes it was.